### ISSUE TITLE
get_grad_reward_function does not return grad for state variables ?

### ISSUE THREAD

**Issue by sbenthall** [scikit-agent/scikit-agent#129]
I'm looking at the automated tests for `get_grad_reward`:

https://github.com/scikit-agent/scikit-agent/blob/main/tests/test_bellman.py#L201

There is a simple block used in the test:

```
# Simple consumption block for testing gradients
        self.simple_block = model.DBlock(
            name="simple_consumption",
            dynamics={
                "c": model.Control(["a"]),
                "a": lambda a, c: a - c,  # Simple wealth dynamics
                "u": lambda c: torch.log(c),  # Log utility
            },
            reward={"u": "consumer"},
        )
```

I would expect `get_grad_reward`, when used `wrt` the state variable `a`, would return a positive, real-valued gradient. Of course, utility increases in wealth.

However, it looks like the current implementation returns `None` for the gradient of `a`:

https://github.com/scikit-agent/scikit-agent/blob/main/tests/test_bellman.py#L266C1-L288C79

The rationale here seems to be that `a` is an "unused variable" in the function that determines `u`. But what we need for the Envelope condition (see #106) is to take the gradient of is not the structural equation that determines the gradient $u = \text{log}(c)$, but rather the gradient of the reward as a function of states, shocks, and control variables $r_s(m, s, x)$ . Likewise, there are models where the relationship between control variables $x$ and reward $r$ is more complicated and so just relying on the structural equation might not work.
